runtime/remote_config.go

```
@@ -0,0 +1,300 @@
+func NewRemoteConfig(cfg *RemoteConfigOptions, logger *zap.Logger, scope tally.Scope) (RemoteConfig, error) {
+	if logger == nil {
+		logger = zap.NewNop()
+	}
+	if len(cfg.FilePath) < 1 || cfg.PollingInterval == 0 {
+		return nil, errors.Errorf("invalid remote config options")
+	}
+	if _, err := os.Stat(cfg.FilePath); os.IsNotExist(err) {
+		return nil, errors.Errorf("invalid remote config file path")
+	}
+	if scope == nil {
+		logger.Warn("no valid metrics scope")
+		scope = tally.NoopScope
+	}
+	rc := &remoteConfig{
+		config:      cfg,
+		close:       make(chan struct{}, 1),
+		subscribers: make(map[string]*func()),
+		logger:      logger,
+		scope:       scope,
+		poller:      time.NewTicker(cfg.PollingInterval),
+	}
+	rc.props.Store(make(RemoteConfigMap))
+	if err := rc.Refresh(); err != nil {
+		return nil, err
+	}
+	rc.wg.Add(1)
+	go rc.poll()
+	return rc, nil
+}
+
+func (rc *remoteConfig) loadConfig() RemoteConfigMap {
+	return rc.props.Load().(RemoteConfigMap)
+}
+
+func (rc *remoteConfig) getValidatedValue(key string, vt jsonparser.ValueType) (*RemoteConfigValue, error) {
+	ret, ok := rc.loadConfig()[key]
+	if !ok {
+		return nil, errors.Errorf("Key <%s> does not exist", key)
+	}
+	if ret.dataType != vt {
+		return nil, errors.Errorf("Key <%s> type mismatch (expected %s actual %s)", key, vt, ret.dataType)
+	}
+	return ret, nil
+}
+
+// GetBoolean returns the value as a boolean
+func (rc *remoteConfig) GetBoolean(key string) (bool, error) {
+	var (
+		ret *RemoteConfigValue
+		err error
+	)
+	if ret, err = rc.getValidatedValue(key, jsonparser.Boolean); err != nil {
+		return false, err
+	}
+	v, _ := jsonparser.ParseBoolean(ret.bytes)
+	return v, nil
+}
+
+// GetFloat returns the value as a float64
+func (rc *remoteConfig) GetFloat(key string, fallback float64) (float64, error) {
+	var (
+		ret *RemoteConfigValue
+		err error
+	)
+	if ret, err = rc.getValidatedValue(key, jsonparser.Number); err != nil {
+		return fallback, err
```

if there is a type mismatch returning the fallback is a good idea.

---

Above 2 comments my take away is we want silent swallow error (?) which I don't think it's a good approach, I get the point that if you pass a fallback value, when fallback happens it's a legit case,  still IMHO caller should decide not remote config

---

If there is a typo in the remote config UI we do not want to hit an err code branch.

Basically if a key does not exist or its configured poorly we should have the same semantics of using the fallback.

We can use metrics/logs/alerts to handle misconfigured keys but our runtime behaviour should just use the fallback.

We should not have code like

```
val, err := rc.GetBool('should.i.do.a', true)
if err != nil {
  return err
}
if val {
  doA();
}
```

In this example the rc key is about whether we should do a thing. If its misconfigured or non-existant we should do the thing. 

We definitely do not want to swallow the error. we can emit logs and metrics.